Support for password decryption from KMS

labianchin opened the pull request:
Contains some tests and docs. Also tested locally and it seems to be working fine.

labianchin commented:
PTAL @honnix @dano as this relates to Styx Secrets.

labianchin commented:
PTAL @honnix @danielnorberg as this relates to Styx Secrets.

honnix commented on README.md:
```diff
@@ -108,6 +108,42 @@ java -cp ./dbeam-core/target/dbeam-core-shaded.jar \
+Database password can be configured by simply passing `--password=writepasswordhere`, `--passwordFile=/path/to/file/containing/password` or `--passwordFile=gs://gcs-bucket/path/to/file/containing/password`.
+
+A more robust configuration is to point to a [Google KMS](https://cloud.google.com/kms/) encrypted file.
+DBeam will try to decrypt using KMS if the file ends with `.encrypted` (e.g. `--passwordFile=gs://gcs-bucket/path/to/db-password.encrypted`).
```
Maybe less automagic by introducing `--passwordFileEncrypted`? Not sure, up to you to decide.

labianchin replied:
Good idea!

honnix commented on README.md:
````diff
@@ -108,6 +108,42 @@ java -cp ./dbeam-core/target/dbeam-core-shaded.jar \
+
+```sh
+java \
+  -PKMS_DEFAULT_KEYRING=dbeam \
````
`-D`?

honnix replied:
Using properties does not work well with docker images such as distroless where entrypoint is `java -jar`. It might be useful to read from properties first, then env and finally default value.

labianchin replied:
Right. For this problem this is not a big problem. When using from docker, another option is to have `JAVA_OPTS` env var, similar to how sbt-pack creates entrypoint scripts

honnix commented on dbeam-core/src/main/java/com/spotify/dbeam/options/PasswordReader.java:
```diff
@@ -37,21 +37,36 @@
+  private final KmsSecrets.Decrypter decrypter;
 
-  public static Optional<String> readPassword(DBeamPipelineOptions options) throws IOException {
+  public PasswordReader(KmsSecrets.Decrypter decrypter) {
```
Is it intended to be `public`?

honnix commented on dbeam-core/src/main/java/com/spotify/dbeam/options/KmsSecrets.java:
```diff
@@ -0,0 +1,185 @@
+public class KmsSecrets {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PasswordReader.class);
+  private static final String DEFAULT_KEYRING;
```
`DEFAULT_` might be misleading because they are the configured values.

honnix commented on dbeam-core/src/main/java/com/spotify/dbeam/options/KmsSecrets.java:
```diff
@@ -0,0 +1,185 @@
+        .location(DEFAULT_LOCATION)
+        .key(DEFAULT_KEY)
+        .keyring(DEFAULT_KEYRING)
+        .project(ServiceOptions.getDefaultProjectId());
```
I remember this has some unexpected behaviour. We might want to make it configuration as well.

honnix commented on dbeam-core/src/main/java/com/spotify/dbeam/options/KmsSecrets.java:
```diff
@@ -0,0 +1,185 @@
+      return decrypter.credentials(GoogleCredential.getApplicationDefault());
+    } catch (IOException e) {
+      LOGGER.warn("Unable to fetch default Google application credential");
+      return decrypter;
```
It might be better to fail loudly instead of returning an unusable decrypter.

labianchin commented:
Thanks for the comments @honnix . I addressed some of the comments.

I also moved `GoogleCredential.getApplicationDefault()` to outside the constructor. Usage of GCP credentials is not a requirement for DBeam.

labianchin commented:
Replaced by #45
